### DISCUSSION TITLE
How does WinPython find and use winpython.ini

### DISCUSSION THREAD

**Discussion by stevecj** [winpython/winpython#1423] [Q&A] [answered]
I am trying to write a compiled wrapper that runs Python programs via WinPython, passing an environment similar to the environment that is set up by `python.bat` (which invokes `env.bat`). I can tell that I'm getting a different result because if I add an entry to the `[environment]` section of `winpython.ini`, then Python can read that value when executed through python.bat but not when executed through my compiled wrapper.

I have tried comparing the environments generated by my wrapper and by `python.bat`, and there is nothing different that seems like it could be relevant.

What am I missing?

Things I have double checked:
* `sys.executable` does have the correct value.
* Values passed into the environment of the subprocess by my executable wrapper are accessible by the Python code.

Why it matters for me:

I don't actually need to configure anything through `winpython.ini`, but the fact that the configuration is not being applied makes me concerned about what else might not work correctly for me using this process.

**Comment by stevecj** [comment, marked as answer]
I finally figured out that `winpython.ini` is actually handled in `env_for_icons.bat`, not env.bat` and that it is converted into commands that are executed from within that batch file that do things like setting environment variables, so that's why nothing in `winpython.ini` was having any effect.

**Comment by stonebig** [comment]
Indeed, this is this beautiful line:

```
FOR /F "delims=" %%i IN ('""%WINPYDIR%\python.exe" "%~dp0WinpythonIni.py""') DO set winpythontoexec=%%i

%winpythontoexec%set winpythontoexec=
```

As you can't propagate upward a change of variables, you have to "replay" the output kindly generated per WinpythonIni.py
(or WinpythonIni.vbs in earlier versions)

**Comment by RoyiAvital** [comment]
Where did the functionality of `winpython.ini` moved onto?
Is the user expected to use `env.ini` in `scripts` instead of `winpython.ini` in `settings`?

**Reply by stonebig** [reply]
Env.ini is to not have to regenerate the env.bat every release . Whatever small things that can trouble an anti virus

**Reply by RoyiAvital** [reply]
So why is `winpython.ini` missing from the latest release?

**Comment by stonebig** [comment]
Hu ? Hummm, did I forgot to force generate it at first icon click ?

**Comment by stonebig** [comment]
apparently i did remove the "useless otherwise" call to get that winpython.ini generated per winpythonini.py.

so until you try an icon, it's not created

... someone has to find where that overkill broomstick gesture happened

**Comment by stonebig** [comment]
before summer 2024 settings/winpython.ini was created if empty per env.bat
now it's only created per env_for_icons.bat via winpython.ini
... may need a tweak of build process

**Comment by stonebig** [comment]
https://github.com/winpython/winpython/issues/1581
